### PR TITLE
[Veeam] Block operations in restoring VMs

### PR DESCRIPTION
**Pull request by SadiJr**
### Description

ACS allows users to modify the target VM when restoring one VM from a backup (VMware + Veeam). This behavior can cause the restore to fail in Veeam and also generate some inconsistencies in the database. This behavior has been fixed, so operations that modify the VM cannot be applied when restoring VMs.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?
It was tested in a local lab:

1. I created one VM, added this VM to one backup offering, and made one manual backup;
2. I tested restoring this backup and tried to destroy the VM during this process;
3. Previously, I could easily destroy the VM;
4. Now, the ACS does not allow me to modify the VM.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test rocky8 vmware-67u3

File: server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java [hunk 1]
```diff
@@ -585,17 +595,64 @@ public boolean restoreBackup(final Long backupId) {
...
+            final BackupProvider backupProvider = getBackupProvider(offering.getProvider());
+            if (!backupProvider.restoreVMFromBackup(vm, backup)) {
+                throw new CloudRuntimeException(String.format("Error restoring %s from backup [%s].", vm, backupDetailsInMessage));
+            }
+        } catch (Exception e) {
+            LOG.error(String.format("Failed to restore backup [%s] due to: [%s].", backupDetailsInMessage, e.getMessage()), e);
+            updateVolumeState(vm, Volume.Event.RestoreFailed);
+            updateVmState(vm, VirtualMachine.Event.RestoringFailed);
+            throw new CloudRuntimeException(String.format("Error restoring VM from backup [%s].", backupDetailsInMessage));
         }
```
Comment (DaanHoogland) [inline server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:614]: can you extract this to a method (end test it)?

File: server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java [hunk 2]
```diff
@@ -585,17 +595,64 @@ public boolean restoreBackup(final Long backupId) {
...
+                LOG.debug(String.format("Trying to update state of volume [%s] with event [%s].", volume, event));
+                try {
+                    if (!volumeApiService.stateTransitTo(volume, event)) {
+                        throw new CloudRuntimeException(String.format("Unable to change state of volume [%s] to [%s].", volume, Volume.State.Restoring));
+                    }
+                } catch (NoTransitionException e) {
+                    String errMsg = String.format("Failed to update state of volume [%s] with event [%s] due to [%s].", volume, event, e.getMessage());
+                    LOG.error(errMsg, e);
+                    throw new RuntimeException(errMsg);
+                }
```
Comment (DaanHoogland) [inline server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:650]: can you extract this to a method?

File: server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java [hunk 3]
```diff
@@ -585,17 +595,64 @@ public boolean restoreBackup(final Long backupId) {
...
-            throw new CloudRuntimeException("Error restoring VM from backup ID " + backup.getId());
+        String backupDetailsInMessage = ReflectionToStringBuilderUtils.reflectOnlySelectedFields(backup, "uuid", "externalId", "vmId", "type", "status", "date");
+        try {
+            updateVmState(vm, VirtualMachine.Event.RestoringRequested);
+            updateVolumeState(vm, Volume.Event.RestoreRequested);
+            final BackupProvider backupProvider = getBackupProvider(offering.getProvider());
+            if (!backupProvider.restoreVMFromBackup(vm, backup)) {
+                throw new CloudRuntimeException(String.format("Error restoring %s from backup [%s].", vm, backupDetailsInMessage));
+            }
+        } catch (Exception e) {
```
Comment (DaanHoogland) [inline server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:609]: if this catch-all is really necessary, please comment on that.

**Review by DaanHoogland — APPROVED** [review]
code looks mostly good, some comments

File: server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java (hunk 1 above)
Comment (SadiJr) [reply server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java]: Done, thanks for the suggestion.

File: server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java (hunk 2 above)
Comment (SadiJr) [reply server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java]: Done, thanks for the suggestion.

File: server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java (hunk 3 above)
Comment (SadiJr) [reply server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:625]: @DaanHoogland has the restore process is made by an external provider, I chose to put a catch-all to ensure that no provider-side exception could be missed. Please let me know if my explanation wasn't enough (communication is not my strong point :()

Comment (DaanHoogland) [reply server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:625]: Your explanation is perfect, @SadiJr better than any I have ever heard about a pokemon catch. I would like to see it in a comment in the code, so people that don't see this PR know the reason as well ;)

Comment (SadiJr) [reply server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:627]: Done, thanks for this suggestions, I hadn't thought of that possibility.

**Comment by DaanHoogland** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=apache_cloudstack&pullRequest=7238)
> 
> [Bug](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=7238&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=7238&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_cloudstack&pullRequest=7238&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=CODE_SMELL) [25 Code Smells](https://sonarcloud.io/project/issues?id=apache_cloudstack&pullRequest=7238&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=7238&metric=new_coverage&view=list) [47.6% Coverage](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=7238&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=7238&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=apache_cloudstack&pullRequest=7238&metric=new_duplicated_lines_density&view=list)

please look at those. I know a lot of useless ones but the ones about `RuntimeException` and the ones about the test case seem valid concerns.

**Review by stephankruggg — APPROVED** [review]
CLGTM, not manually tested

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
insanity is
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
trusting testing by the user, merging

**Pull request merged**